### PR TITLE
CFIN 402 Create new style tags

### PR DESCRIPTION
**Pull request by ncn1**
For the purpose of being able to demo the styling etc, each course has 2 hard-coded tags for the extra details such as year abroad etc that are not currently being passed in the Courses API.

This has also necessitated some creativity with the tests, which I will comment on in situ

### CODE REVIEW

File: src/components/CourseDetails.js [hunk 1]
```diff
@@ -1,15 +1,31 @@
...
 const CourseDetails = ({ course }) => {
     return (
         <TagList>
-            <AwardTag award={course.award} />
-            <StartDateTag yearOfEntry={course.yearOfEntry} />
-            <LengthCourseTag length={course.length} />
+            <LevelAndAwardTag level={course.level} award={course.award} />
+            <StartDateAndLengthTag yearOfEntry={course.yearOfEntry} length={course.length} />
+            <ExtrasTag
+                title="Extras"
```
Comment (ncn1) [inline src/components/CourseDetails.js:13]: Hard-coded tags

File: src/tests/components/CourseDetails.test.js [hunk 2]
```diff
@@ -12,43 +13,123 @@ describe("Course details", () => {
...
+        const tag = screen.queryAllByRole("listitem");
+
+        if (tag.length === 0) {
+            expect((tag.length = 0));
+        } else {
+            expect(tag[0]).not.toHaveTextContent("undergraduate");
+            expect(tag[0]).not.toHaveTextContent("BA (Hons)");
+            expect(within(tag[0]).queryByTestId("tag-icon")).toBe(null);
+        }
+    });
```
Comment (ncn1) [inline src/tests/components/CourseDetails.test.js:75]: This is the fiddle to ensure the test works with the hard-coded tags, in which case there will be some tags returned, and also when there are no hard-coded tags, so queryAllByRole("listitem") returns an empty array. Will need refactoring once design is finalised.

**Review by mrgibbonsuoy — APPROVED** [review]

File: src/components/CourseDetails.js [hunk 3]
```diff
@@ -1,15 +1,31 @@
...
+                    { icon: "plane", text: "year in industry" },
+                    { icon: "mortar-board", text: "Distance Learning" },
+                ]}
+            />
+            <ExtrasTag
+                title=""
+                content={[
+                    { icon: "suitcase", text: "year abroad" },
+                    { icon: "plane", text: "year in industry" },
+                    { icon: "plane", text: "Distance Learning" },
```
Comment (FergusMcGlynn) [inline src/components/CourseDetails.js:25]: Should the "Distance Learning" icon in this tag be consistent with the "Distance Learning" icon in the previous `ExtrasTag`?

File: src/components/CourseDetails.js (hunk 1 above)
Comment (FergusMcGlynn) [reply src/components/CourseDetails.js:13]: Are these hard coded in order to demonstrate how they look? Presumably we don't want them present on every course result?

File: src/components/CourseDetails.js [hunk 4]
```diff
@@ -18,40 +34,27 @@ CourseDetails.propTypes = {
...
 };
 
-const AwardTag = ({ award }) => {
-    if (award) {
-        return <Tag icon="university" mainText={award} />;
-    }
-
-    return null;
+const LevelAndAwardTag = ({ level, award }) => {
+    const tag = level || award ? <Tag topIcon="mortar-board" title={level} subText={award} /> : null;
```
Comment (FergusMcGlynn) [inline src/components/CourseDetails.js:38]: I find this line a little hard to parse - could we expand it to make it easier to read?

File: src/components/Tag.js [hunk 5]
```diff
@@ -1,18 +1,53 @@
...
 
-const Tag = ({ icon, mainText }) => (
-    <li className="c-tag-list__item">
-        <span className="c-tag">
-            <i aria-hidden className={`c-tag__icon c-icon c-icon--${icon}`} data-testid="tag-icon" />
-            {mainText}
-        </span>{" "}
-    </li>
+const Tag = ({ topIcon, title, subText }) => (
+    <div className="new-tag" role="listitem">
```
Comment (FergusMcGlynn) [inline src/components/Tag.js:5]: Is the class name `new-tag` for illustrative purposes? Fine if so, but if not I think it could be confusing.

File: src/pages/_app.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+const MyApp = ({ Component, pageProps }) => {
+    return <Component {...pageProps} />;
+};
+
+MyApp.propTypes = {
+    Component: PropTypes.any,
+    pageProps: PropTypes.any,
+};
+
+export default MyApp;
```
Comment (FergusMcGlynn) [inline src/pages/_app.js:14]: This is new to me - how does this relate to the new tags design?

Comment (FergusMcGlynn) [reply src/pages/_app.js:14]: Also is `MyApp` the most appropriate name for this function? (It may well be - I don't understand the context enough to know - but it doesn't seem like a very descriptive name)

File: src/components/CourseDetails.js (hunk 3 above)
Comment (ncn1) [reply src/components/CourseDetails.js:25]: Yes - see introductory comment

File: src/components/CourseDetails.js (hunk 1 above)
Comment (ncn1) [reply src/components/CourseDetails.js:13]: Yes - see introductory comment

Comment (RichardAllum) [reply src/components/CourseDetails.js:13]: They're just there for demoing, yeah. This PR isn't going into dev, so they'll be removed before that

File: src/components/CourseDetails.js (hunk 3 above)
Comment (RichardAllum) [reply src/components/CourseDetails.js:25]: Maybe! It's entirely seeing if things exist at the moment

Comment (ncn1) [reply src/components/CourseDetails.js:25]: Would be less confusing, even though they are just here for demo purposes. I will change.

File: src/components/CourseDetails.js (hunk 4 above)
Comment (ncn1) [reply src/components/CourseDetails.js:38]: It's basically the equivalent of the Groovy elvis operator - if there is either a level or award, return something, else return nothing. It's a choice of this succinct one -liner or 6 lines of an if / else statement.

File: src/components/CourseDetails.js [hunk 7]
```diff
@@ -18,40 +34,27 @@ CourseDetails.propTypes = {
...
+    return tag;
 };
```
Comment (RichardAllum) [inline src/components/CourseDetails.js:40]:
```suggestion
const LevelAndAwardTag = ({ level, award }) => {
    if (level || award) {
        return <Tag topIcon="mortar-board" title={level} subText={award} />
    }
    return null;
};
```
per Ferg's comment above.

File: src/tests/components/CourseDetails.test.js [hunk 8]
```diff
@@ -12,43 +13,123 @@ describe("Course details", () => {
...
+    it("does not display level and award tag when neither level or award is present", () => {
+        const exampleCourse = {};
+
+        render(<CourseDetails course={exampleCourse} />);
+
+        const tag = screen.queryAllByRole("listitem");
+
+        if (tag.length === 0) {
+            expect((tag.length = 0));
+        } else {
```
Comment (FergusMcGlynn) [inline src/tests/components/CourseDetails.test.js:70]: I don't understand this bit. Presumably `expect((tag.length = 0));` is doing nothing because it's assigning `0` to `tag.length`, and that's not something I'd expect to verify with `expect`. Can this be simplified to:
```suggestion
        if (tag.length > 0) {
```
(I don't know whether it can BTW)

Aside from the above - I think conditional checks in tests are a bit strange and I think we should avoid them. As you say, this is going to be refactored at some point. Although I wouldn't want it to be merged in to `dev` like this if we can avoid it.

**Review by RichardAllum — CHANGES_REQUESTED** [review]
I think we'll want to take the `Tag/ArrayTag` components and commit them to CFIN-371 in a brand new commit once the styles are in the pattern library

File: src/components/CourseDetails.js (hunk 7 above)
Comment (ncn1) [reply src/components/CourseDetails.js:40]: I think as we use the original style a bit more, it will become familiar and no longer difficult to parse. When learning Groovy, I found most of it's shortcuts difficult to read, but came to really like them as the language became familiar.

File: src/pages/_app.js (hunk 6 above)
Comment (ncn1) [reply src/pages/_app.js:14]: It's code that needs to be included so you can include a local css file. I blindly followed some instructions, so I don't actually understand it myself. Hopefully we can remove it once the new styles are in the pattern library.

File: src/tests/components/CourseSearchResults.test.js [hunk 9]
```diff
@@ -10,8 +10,10 @@ describe("CourseSearchResults", () => {
 
         render(<CourseSearchResults isSuccessfulSearch searchResults={searchResults} />);
 
-        expect(screen.getByRole("link", { name: "Maths" })).toBeInTheDocument();
-        expect(screen.getByRole("link", { name: "Physics" })).toBeInTheDocument();
+        const headings = screen.getAllByRole("heading");
+
+        expect(headings[0]).toHaveTextContent("Maths");
+        expect(headings[1]).toHaveTextContent("Physics");
```
Comment (FergusMcGlynn) [inline src/tests/components/CourseSearchResults.test.js:16]: Any particular reason for selecting these as headings as opposed to links? I appreciate that they are both, but from a user's perspective I would have thought it's more important that they are links than that they are headings.

File: src/tests/components/TagList.test.js [hunk 10]
```diff
@@ -1,23 +1,71 @@
...
+        expect(tags[0]).toHaveTextContent("year in industry");
+        expect(within(tags[0]).getByTestId("suitcase")).toHaveClass("c-icon--suitcase");
+        expect(within(tags[0]).getByTestId("plane")).toHaveClass("c-icon--plane");
+        expect(within(tags[0]).getByTestId("mortar-board")).toHaveClass("c-icon--mortar-board");
+
+        expect(tags[4]).not.toHaveTextContent("Extras");
+        expect(tags[4]).toHaveTextContent("year abroad");
+        expect(tags[4]).toHaveTextContent("year in industry");
+        expect(tags[4]).toHaveTextContent("Distance Learning");
+        expect(tags[4]).toHaveTextContent("Integrated Masters");
```
Comment (FergusMcGlynn) [inline src/tests/components/TagList.test.js:65]: It's a bit weird to the reader how we're jumping to the 5th element in the `tags` array. I'm guessing that's because we have lists within lists, and `getAllByRole('listitem')` is fetching every list item. It makes me wonder whether this would be confusing for a screen reader.

A possible alternative might be to use `<dl>`, `<dt>` and `<dd>` for the sub lists. `<dt>` has the implicit role `term` and `<dd>` has the implict role `definition`, so we'd be able to `getAllByRole('term')` or `getAllByRole('definition')` and that might be less confusing to a screen reader.

File: src/tests/components/TagList.test.js [hunk 11]
```diff
@@ -1,23 +1,71 @@
...
+        expect(within(tags[4]).getByTestId("suitcase")).toHaveClass("c-icon--suitcase");
+        expect(within(tags[4]).getByTestId("mortar-board")).toHaveClass("c-icon--mortar-board");
+        const planes = within(tags[4]).getAllByTestId("plane");
+        expect((planes.length = 2));
```
Comment (FergusMcGlynn) [inline src/tests/components/TagList.test.js:69]: I don't think `plane` is a good name for a test id

File: src/components/CourseDetails.js (hunk 7 above)
Comment (FergusMcGlynn) [reply src/components/CourseDetails.js:40]: It's not clear what the operator precedence is in the original line - does the `||` take precedence, or does the `?`?

Comment (FergusMcGlynn) [reply src/components/CourseDetails.js:40]: I prefer Richard's suggestion. It's not quite equivalent to the Groovy elvis operator because of the `||` - which is what justifies breaking it down.

File: src/tests/components/CourseDetails.test.js [hunk 12]
```diff
@@ -12,43 +13,123 @@ describe("Course details", () => {
...
+        const exampleCourse = {
+            level: "undergraduate",
+        };
+
+        render(<CourseDetails course={exampleCourse} />);
+
+        const tag = screen.getAllByRole("listitem");
+
+        expect(tag[0]).toHaveTextContent("undergraduate");
+        expect(tag[0]).not.toHaveTextContent("BA (Hons)");
```
Comment (FergusMcGlynn) [inline src/tests/components/CourseDetails.test.js:43]: I find this line (`.not.toHaveTextContent("BA (Hons)")`) a bit arbitrary. It might as well say `.not.toHaveTextContent("banana")`. If the goal is to verify that only the text `undergraduate` is presented and no other text whatsoever then I suspect there's a more robust way to verify it.

Comment (FergusMcGlynn) [reply src/tests/components/CourseDetails.test.js:43]: Looks like you can use [a regular expression inside `toHaveTextContent`](https://github.com/testing-library/jest-dom#tohavetextcontent) - that could be a way of verifying that exactly `undergraduate` is matched and there's no other text:

```
expect(tag[0]).toHaveTextContent(/^undergraduate$/);
```

**Review by FergusMcGlynn — CHANGES_REQUESTED** [review]
I'm not quite sure how much of this PR is demo-ing ideas and how much of it is intended to be merged into `dev`, so I feel conflicted about whether or not to `Request changes`. I'll do so, to reflect that I think some tests could be improved, in line with the comments I've left. Apologies if that's not keeping in the spirit of this PR!

**Comment by FergusMcGlynn** [comment]
As an aside - it might be worth marking this PR as a "draft", if its purpose is to showcase some changes without requesting the full scrutiny of a code review. I'm still not entirely clear what level of reviewing is being requested.

File: src/components/CourseDetails.js (hunk 4 above)
Comment (ncn1) [reply src/components/CourseDetails.js:38]: Fergus, can I not persuade you to embrace the terniary operator for single if - then - else 's like this one? https://www.javascripttutorial.net/javascript-ternary-operator/ . I think nesting them gets very hard to follow, but I think single ones do remove a lot of verbosity.

Comment (ncn1) [reply src/components/CourseDetails.js:38]: Sorry, did not see your reply to the other thread before adding my previous comment.

Comment (FergusMcGlynn) [reply src/components/CourseDetails.js:38]: I'm fine with ternary operators for simple `if a then b else c` expressions, but this one is a bit more complicated than that. Is it saying 
```
level || (award ? <Tag ... /> : null)
```
or
```
(level || award) ? <Tag .../> : null
```
? It's not immediately obvious to me. Also I think one of the sub-expressions being JSX makes it a bit trickier to parse than a simple all-javascript ternary. 

Richard suggested an alternative which makes it really easy to parse, and it's only a couple of lines. I'm not going to die in a ditch over it, and if everyone else finds the ternary perfectly easy to understand then we can stick with it.

File: src/components/CourseDetails.js (hunk 7 above)
Comment (ncn1) [reply src/components/CourseDetails.js:40]: Annoyingly, if I put brackets around 'level || award' in the terniary version, Prettier removes them!

File: src/tests/components/CourseDetails.test.js (hunk 12 above)
Comment (ncn1) [reply src/tests/components/CourseDetails.test.js:43]: Thanks - much better way of doing it.

File: src/tests/components/TagList.test.js (hunk 10 above)
Comment (ncn1) [reply src/tests/components/TagList.test.js:65]: Description lists are only supposed to have 1 <dd> element, whereas our Extras tab needs there to be multiple lines under the <dt>. So standard lists it needed to be. I agree it isn't great - I keep coming up against the need for a getByClass option, but this has so far been ruled out by the authors of the testing library.

File: src/tests/components/TagList.test.js (hunk 11 above)
Comment (ncn1) [reply src/tests/components/TagList.test.js:69]: It means it can be the same as the icon name. Again, class name would have been much simpler to use here, and negate the need for a test id.

File: src/tests/components/CourseSearchResults.test.js (hunk 9 above)
Comment (ncn1) [reply src/tests/components/CourseSearchResults.test.js:16]: Yes, the name of the links now also includes all the text in the tags.

Comment (FergusMcGlynn) [reply src/tests/components/CourseSearchResults.test.js:16]: Isn't that potentially an accessibility problem? If a screen reader is going to read out everything for a link? This might need some thought.

Comment (ncn1) [reply src/tests/components/CourseSearchResults.test.js:16]: I think that should be left for CFIN-401 'Ensure accessibility of new tags' to check and address if necessary. I has assumed it was just a React testing library artifact.

Comment (FergusMcGlynn) [reply src/tests/components/CourseSearchResults.test.js:16]: Fair enough - so long as it's not forgotten

**Comment by ncn1** [comment]
> I'm not quite sure how much of this PR is demo-ing ideas and how much of it is intended to be merged into `dev`, so I feel conflicted about whether or not to `Request changes`. I'll do so, to reflect that I think some tests could be improved, in line with the comments I've left. Apologies if that's not keeping in the spirit of this PR!

It was a bit of both, and the code review has been really helpful. But if you are now ok with the changes that have already been made and knowing that it will not go into dev like this (and that would require another pull request, anyway, as this is just merging into the story branch) would you be able to approve it now?

**Review by FergusMcGlynn — APPROVED** [review]

**Comment by ncn1** [comment]
As Richard has pointed out, probably best not to merge these changes into the story branch, as all the hard-coding and tests that account for the hard-coding will need removing before merging into dev, so closing the PR without merging.
